### ISSUE TITLE
Rule proposal: Allow use of arithmetic & logical operators only for strings and numbers

### ISSUE THREAD

**Issue by pbadenski** [typescript-eslint/typescript-eslint#3200]
## Rule Details

This rule restricts the use of arithmetic & logical operators to strings and numbers.

## Fail

```
1 + {}
```
```
+{}
```
```
2 * function() { }
```
```
undefined & new Object()
```
```
var o = {};
o++;
```

## Pass

```
1 + 2
```
```
"a" + 2
```
```
String("a") + Number("1")
```
```
var n = 0;
n++
```

**Comment by bradzacher** [comment]
Have you tried this rule?

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/restrict-plus-operands.md

**Comment by pbadenski** [comment]
Yes, I have - it does cover some of the cases, but not sufficiently enough. For example:

https://yeonjuan.github.io/typescript-eslint-demo/#e2NvZGU6J3ZhciBhID0gdW5kZWZpbmVkICYgbmV3IE9iamVjdCgpO1xudmFyIGIgPSAre307XG52YXIgYyA9IHt9O1xuYysrO1xuXG5cbicscnVsZXM6eydAdHlwZXNjcmlwdC1lc2xpbnQvcmVzdHJpY3QtcGx1cy1vcGVyYW5kcyc6J2Vycm9yJ30scGFyc2VyT3B0aW9uczp7ZWNtYUZlYXR1cmVzOntqc3g6ZmFsc2UsZ2xvYmFsUmV0dXJuOmZhbHNlfSxlY21hVmVyc2lvbjoyMDE4LHByb2plY3Q6WycuL3RzY29uZmlnLmpzb24nXSxzb3VyY2VUeXBlOidzY3JpcHQnfX0=

**Comment by bradzacher** [comment]
The reason the rule doesn't report is because hte majority of your examples are TS errors:

```ts
1 + {}; // Operator '+' cannot be applied to types 'number' and '{}'.(2365)

2 * function() { }; // The right-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.(2363)

undefined & new Object(); // Object is possibly 'undefined'.(2532)
                          // The right-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.(2363)

({})++; // An arithmetic operand must be of type 'any', 'number', 'bigint' or an enum type.(2356)

+{}; // no TS error

```

[repl](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=12&pc=1#code/IwAg1CDeC+DcIHoEgPIAcCmAnAhgFwHssQByMEkAYxwDsaC8QAjDEHNNAGwEsMATEIUEBPTAGdSNAK4BbFlgq0BJGCQB0ACgBMAZgBsAVgCUAKBNaQAKhAAzKTUp5uBGhqNQQcRMgAqAC1YsbgBzPzwAWj8lEDFuPlYCGzYaNiC8PxkMJ0oQAkxcJxcQGSkxRhZcpLxRVhJaYRIAGklZeSbSJhDuGjwKImSQDGkZEUxNXT0dUxN7eJtu-hAAMhAaDAB3VCYAKwxHN3gkLd3HEG4JNAIxWKZOYVJZjHm1vnVtAx0tUxAf37--gH-I7+QIhMKRaKxeKVAY4NIZLLcHJ5bD4ZwpEplZgJKo1Uj1dokYZtZokTrBbq9XLEWiDYajDDjfRTMwaGBGMBgQ7IACCKTh3HSmWyuXy0Ux5RxDPxNAapOJ2EJ5MpfRpKSGsgZTIMemmYBg3NWBBAPgAyoMsFgiGYgA)

Making a lint rule duplicate TS errors is something we *do not* do - on purpose.

The only case that's not handled is `+{}`, which I'll be honest when I say is a suuuuper rare case - I don't know if I've ever seen someone do numerical conversion via a unary `+`.

I'd say that'd be worth raising as a feature request for TS itself, given that it supports most other cases of using objects in arithmetic.

**Comment by pbadenski** [comment]
Ha, apologies.

It's the unary operator that was an issue for us - we use it for numerical conversion in the project. I then listed all other examples without verifying it (embarassing!).

I will take the unary operator type usage verification with TS project, thank you!
